File: Carts/Controllers/Cart Item View Controller/CAListTableViewController.m
//
//  CATableViewController.m
//  GroceryList
//
//  Created by Hakon Hanesand on 1/18/15.

#import <ReactiveCocoa/ReactiveCocoa.h>

#import "CAListTableViewController.h"
#import "CATableViewCell.h"
#import "CAScannerViewController.h"
#import "CAListObject.h"
#import "CABarcodeObject.h"

#import "PFQueryTableViewController+Caching.h"
#import "PFObject+CAPFObject.h"
#import "CAPullToCloseTransitionManager.h"
#import "UIImageView+AFNetworking.h"
#import "CAListItemObject.h"
#import "RACUnit.h"
#import "PFQuery+CAQuery.h"
#import "PFUser+CAUser.h"

static NSString *const kCATableViewReuseIdentifier = @"CATableViewCellIdentifier";
static NSString *const kCAParsePinName = @"CATableViewPin";

@interface CAListTableViewController ()
@property (nonatomic) CAScannerViewController *scanner;
@end

@implementation CAListTableViewController

- (instancetype)initWithStyle:(UITableViewStyle)style {
    if (self = [super initWithStyle:style]) {
        self.pullToRefreshEnabled = YES;
        self.paginationEnabled = NO;
        self.loadingViewEnabled = NO;
        
        self.scanner = [CAScannerViewController instance];
        self.scanner.view.frame = CGRectMake(0, 0, CGRectGetWidth([UIScreen mainScreen].bounds), CGRectGetHeight([UIScreen mainScreen].bounds));
    }
    
    return self;
}

#pragma mark - View Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(didPressAddButton)];
    UIBarButtonItem *flexibleSpace = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
    [self setToolbarItems:@[flexibleSpace, button, flexibleSpace]];
    
    [self.tableView registerNib:[UINib nibWithNibName:NSStringFromClass([CATableViewCell class]) bundle:nil] forCellReuseIdentifier:kCATableViewReuseIdentifier];
    
    [self subscribeToScannerSignal];
}

- (void)subscribeToScannerSignal {
    [[[self.scanner.listItemSignal takeUntil:self.rac_willDeallocSignal] deliverOnMainThread] subscribeNext:^(CABarcodeObject *barcodeObject) {
        CAListItemObject *object = [self.user.list.items.rac_sequence objectPassingTest:^BOOL(CAListItemObject *object) {
            return [object.item isEqual:barcodeObject];
        }];
        
        if (object) {
            object.quantity = [NSNumber numberWithInt:[object.quantity intValue] + 1];
        } else {
            [self.user.list.items addObject:[CAListItemObject objectWithBarcodeObject:barcodeObject]];
        }
        
        [[self.user pinWithSignal] subscribeCompleted:^{
            [self.user saveInBackground];
            [self loadObjects];
        }];
    }];
}

- (void)didPressAddButton {
    [self presentViewController:self.scanner animated:YES completion:nil];
}

- (BOOL)prefersStatusBarHidden {
    return NO;
}

#pragma mark - Parse

- (RACSignal *)signalForTable {
    PFQuery *query = [CAListObject query];
    [query includeKey:@"items.item"];
    
    return [[query getObjectWithIdWithSignal:[PFUser currentUser].list.objectId] map:^NSArray *(CAListObject *listObject) {
        return listObject.items;
    }];
}

- (RACSignal *)cachedSignalForTable {
    PFQuery *query = [CAListObject query];
    [query includeKey:@"items.item"];
    [query fromLocalDatastore];
    
    return [[query getObjectWithIdWithSignal:[PFUser currentUser].list.objectId] map:^NSArray *(CAListObject *listObject) {
        return listObject.items;
    }];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath object:(CAListItemObject *)object {
    CATableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCATableViewReuseIdentifier forIndexPath:indexPath];
    
    cell.name.text = object.item.name;
    cell.brand.text = object.item.brand;
    cell.category.text = object.item.category;
    
    [cell.image setImageWithURL:[NSURL URLWithString:object.item.image[0]]];
    
    return cell;
}

#pragma mark - Tableview data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if ([self.objects count] == 0) {
        //notify the user that there are no saved items
        UILabel *noSavedBarcodesLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.view.bounds))];
        noSavedBarcodesLabel.text = @"You have not saved any barcodes.";
        noSavedBarcodesLabel.textColor = [UIColor blackColor];
        noSavedBarcodesLabel.numberOfLines = 0;
        noSavedBarcodesLabel.textAlignment = NSTextAlignmentCenter;
        noSavedBarcodesLabel.font = [UIFont fontWithName:@"Palatino-Italic" size:15];
        [noSavedBarcodesLabel sizeToFit];
        
        self.tableView.backgroundView = noSavedBarcodesLabel;
        self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
        
        return 0;
    } else {
        self.tableView.backgroundView = nil;
        self.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
        return 1;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 71;
}

- (void)setUser:(PFUser *)user {
    _user = user;
    self.title = [self.user bestName];
}

@end

File: Carts/Controllers/Cart Item View Controller/CAListTableViewController.swift
//
//  CAListTableViewController.swift
//  Carts
//
//  Created by Hakon Hanesand on 4/26/15.
//  Copyright (c) 2015 Hakon Hanesand. All rights reserved.
//

import UIKit

class CAListTableViewController: CAQueryTableViewController, UITableViewDelegate {

    var scanner: CAScannerViewController
    var user: PFUser? {
        didSet {
            self.title = self.user!.bestName()
        }
    }
    
    required init(style: UITableViewStyle) {
        scanner = CAScannerViewController.instance()
//         self.scanner.view.frame = CGRectMake(0, 0, CGRectGetWidth([UIScreen mainScreen].bounds), CGRectGetHeight([UIScreen mainScreen].bounds));
        
        super.init(style: .Plain, className: "")
        
        self.pullToRefreshEnabled = true
        self.paginationEnabled = false
        self.loadingViewEnabled = false
    }

    required init!(coder aDecoder: NSCoder!) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let barButton = UIBarButtonItem(barButtonSystemItem: .Add, target: self, action: "didPressAddButton")
        let flexibleSpace = UIBarButtonItem(barButtonSystemItem: .FlexibleSpace, target: nil, action: nil)
        self.setToolbarItems([flexibleSpace, barButton, flexibleSpace], animated: false)
        
        self.tableView.registerNib(UINib(nibName: "CATableViewCell", bundle: nil), forCellReuseIdentifier: "CATableViewCellIdentifier")
        
        self.subscribeToScannerSignal()
    }
    
    private func subscribeToScannerSignal() {
        self.scanner.listItemSignal?.takeUntil(self.rac_willDeallocSignal()).deliverOnMainThread().subscribeNext({ (next: AnyObject!) -> Void in
            let barcodeObject = next as! CABarcodeObject
            
            let existing = self.list().items.rac_sequence.objectPassingTest({ (next: AnyObject!) -> Bool in
                let object = next as! CAListItemObject
                return object.item.isEqual(barcodeObject)
            }) as? CAListItemObject
            
            if existing == nil {
                self.list().items.addObject(CAListItemObject(barcodeObject: barcodeObject))
            } else {
                existing!.quantity = Int(existing!.quantity) + 1
            }
        })
    }
    
    func list() -> CAListObject {
        return self.user!.valueForKey("list") as! CAListObject
    }
    
    func didPressAddButton() {
        self.presentViewController(self.scanner, animated: true, completion: nil)
    }
    
    override func prefersStatusBarHidden() -> Bool {
        return false
    }
    
    override func signalForTable() -> RACSignal! {
        let query = CAListObject.query()!
        query.includeKey("items.item")
        
        return query.getObjectWithIdWithSignal(PFUser.currentUser()!.list().objectId!).map {
            if let listObject = $0 as? CAListObject {
                return listObject.items
            }
            
            return nil
        }
    }
    
    override func cachedSignalForTable() -> RACSignal! {
        let query = CAListObject.query()!.includeKey("items.item").fromLocalDatastore()
        
        return query.getObjectWithIdWithSignal(PFUser.currentUser()!.list().objectId!).map {
            if let listObject = $0 as? CAListObject {
                return listObject.items
            }
            
            return nil
        }
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("CATableViewCellIdentifier", forIndexPath: indexPath) as! CATableViewCell
        
        if let listItem = self.objectAtIndexPath(indexPath) as? CAListItemObject {
            cell.name.text = listItem.item.name
            cell.brand.text = listItem.item.brand
            cell.brand.text = listItem.item.category
            
            let url = NSURL(string: listItem.item.image[0] as! String)!
            let request = NSMutableURLRequest(URL: url)
            
            cell.image.setImageWithURLRequest(request, placeholderImage: nil, success: { (request, response, image) -> Void in
                cell.image.image = image;
                cell.image.setNeedsDisplay()
                }, failure: nil)
        }
        
        return cell
    }
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        if self.objects!.count == 0 {
            let noSavedBarcodesLabel = UILabel(frame:CGRectMake(0, 0, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.view.bounds)))
            noSavedBarcodesLabel.text = "You have not saved any barcodes."
            noSavedBarcodesLabel.textColor = UIColor.blackColor()
            noSavedBarcodesLabel.numberOfLines = 0
            noSavedBarcodesLabel.textAlignment = .Center
            noSavedBarcodesLabel.font = UIFont(name: "Palatino-Italic", size: 15)
            noSavedBarcodesLabel.sizeToFit()
            
            self.tableView.backgroundView = noSavedBarcodesLabel
            self.tableView.separatorStyle = .None
            
            return 0
        } else {
            self.tableView.backgroundView = nil
            self.tableView.separatorStyle = .SingleLine
            return 1
        }
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 71
    }
}
